### ISSUE TITLE
Issue with nginx Basic Auth

### ISSUE THREAD

**Issue by patsa** [mealie-recipes/mealie#690]
Hello,

I installed mealie using docker-compose on my server. My docker-compose.yml looks as follows:

```
version: "3.1"
services:
  mealie:
    container_name: mealie
    image: hkotel/mealie:latest
    restart: always
    ports:
      - 127.0.0.1:9925:80
    environment:
      PUID: 1000
      PGID: 1000
      TZ: Europe/Berlin

      # Default Recipe Settings
      RECIPE_PUBLIC: 'true'
      RECIPE_SHOW_NUTRITION: 'true'
      RECIPE_SHOW_ASSETS: 'true'
      RECIPE_LANDSCAPE_VIEW: 'true'
      RECIPE_DISABLE_COMMENTS: 'false'
      RECIPE_DISABLE_AMOUNT: 'false'

      DEFAULT_EMAIL: myemail
      API_DOCS: 'false'
      TOKEN_TIME: 24
      # Gunicorn
      WEB_CONCURRENCY: 2
      # WORKERS_PER_CORE: 0.5
      # MAX_WORKERS: 8
    volumes:
      - ./mealie/data/:/app/data
```

So far so good. It runs as expected on port 9925. Although it is important to mention, that without the 127.0.0.1: before 9925, docker has the unpleasant habit of exposing the port to the outside world even though services like ufw should stop this behaviour. As it turns out, docker overwrites the iptables and bypasses ufw by doing so. However by writing the host in front of the local port, the port stays closed. At least to my knowledge. 

But now to my problem:

As I didn't want to fiddle around with SWAG (my server already has nginx up and running), I created a conf file for mealie, in which I use a reverse proxy to direct calls to my domain to the mealie container:

```
server {

    server_name example.com;

    access_log /var/log/... combined buffer=64k flush=5m if=$loggable;
    error_log /var/log/...;

    location / {
       proxy_set_header Authorization "";
       auth_basic            "Basic Auth";
       auth_basic_user_file  "/location/of/my/password/file";
       proxy_pass http://localhost:9925;
       proxy_set_header X-Real-IP $remote_addr;
       proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
       proxy_set_header X-Forwarded-Proto $scheme;
       proxy_set_header Accept-Encoding "";
       proxy_read_timeout 90;
    }

    ... 
    # some more Let's Encrypt stuff
```

The basic auth works like a charm for the frontend and the normal statically served elements. But as soon as an URL with the word 'api' in it is requested, it returns 401, as it is obviously also trying to use basic auth, but can't get any credentials. 

Even though this is only partially related to Mealie itself, I'm sure I'm not the only one interested in a solution to this problem, as the general authentication layer is only supposed to come around version 1.0 and I don't want everyone to be able to see my recipes in the WWW.

Does anyone have any idea how to solve the problem? I guess it will involve changing the Caddyfile in some way.

tldr; When using native nginx with basic authentication, all API requests return error 401.

Thanks in advance to everyone!

Best regards, 
Patrick

**Comment by hay-kot** [comment]
I don't believe this is a problem with Caddy. We're using Axios on the front end to do the API requests on the client, as such you'd have to inject those BasicAuth credentials into the Axios header somehow in the event that BasicAuth is being used.

I have not looked into this, but I imagine implementing some sort of check and header settings would be easy to do...however I've been wrong before. Open to PRs on this.

**Comment by dnhckt** [comment]
Maybe this way would work?:

    location / {
                         location /api/ {
                         
                                   auth_basic      off;
                                   
                               <-- proxy_pass / proxy_set_header stuff goes here -- >    
                                 }
                                          
       auth_basic            "Basic Auth";
       auth_basic_user_file  "/location/of/my/password/file";
       
       <-- proxy_pass / proxy_set_header stuff goes here -- > 
    }
    
I tried this on my instance and it asks for the .htpasswd on the homepage , but once entered all recipes look to come through as normal

**Comment by patsa** [comment]
@dnhckt 
Thank you for your input!
It actually does work a bit like that, but still has its flaws. I can reach the site and use it by entering the htpasswd once. But as soon as I reload, I get asked for the password again.
Also, I can login and can see the edit user information screen, but I can't edit them. The API call still returns 401 Unauthorized. When I then reload the page, the login state of Mealie itself is also lost again and I have to login again.

Any other input on how to achieve a more secure environment to host mealie?

**Comment by Christoph-87** [comment]
Hi all,
I am not sure about the answer above. Can you simply try the following?
location / { -- basic auth here -- }
location /api { }

Using the following URL to verify it, it should work. https://detailyang.github.io/nginx-location-match-visible/

**Comment by DerfOh** [comment]
Disreguard I'm not using basic auth on the middleware level.

~~Here's what I'm using,~~

```
server {

	listen 443;

	root /var/www;
	index index.php index.html index.htm index.nginx-debian.html;

	server_name "mealie.home.lan";

	include /etc/nginx/sites-dietpi/*.conf;

	ssl                  on;
	ssl_certificate      /etc/nginx/ssl/home.lan-wildcard.crt;
	ssl_certificate_key  /etc/nginx/ssl/home.lan-wildcard.key;
	ssl_session_timeout  5m;

	location / {
		proxy_set_header X-Forwarded-Proto https;
		proxy_set_header X-Forwarded-Host $host:$server_port;
        	proxy_set_header X-Forwarded-Server $host;
        	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_pass http://mealie.lan:9925;
	}

}
```

**Comment by fleaz** [comment]
Yes, this is indeed not a problem with any of the involved webservers. The problem is just like hay-kot said the HTTP calls made from within the JS code. Entering the BasicAuth credentials when you load the page, will result in your browser caching and sending them every time you visit a page manually, but the calls to the API are technically not made by your browser itself but by the Javascript which is not aware of any BasicAuth stuff going on, there one would need to add the Authorization header to every axios call (We had exactly this problem a while back at work therefore I immediately recognized it :D )

Disabling BasicAuth on the whole `/api/` path is a bad recommendation imho and just security by obscurity.

If you just don't want anybody to see your recipes, you could just set all recipes to private via the `RECIPE_PUBLIC` environment variable, then every visitor will just see a blank Mealie installation.


Or just wait for v1.0 which will allow only logged in users to access your installation ;)

**Comment by tdharris** [comment]
Still an issue I think. It appears I can't secure the reverse-proxy route in nginx with basic auth. Only thing that seems to work is disabling basic auth completely for the route, which is not something I particularly like doing as it means many of the assets are loaded unprotected. From a security perspective, it's best to prevent the upstream connection entirely in this case.

Though, I see am using `latest` docker tag from [hkotel/mealie](https://hub.docker.com/r/hkotel/mealie), which appears to be `v0.5.6` so if fixed since then, sorry for the trouble.

---

**EDIT**: I've since switched to Authelia, which helps to avoid the problem for me.
